language/vm/src/binary_format.rs

```diff
@@ -0,0 +1,362 @@
+    // specifications are references to a *generic* handle and and its instantiation, whether
+    // partial or complete instantiation
+    pub struct_specs: Vec<StructSpec>,
+    pub function_specs: Vec<FunctionSpec>,
+    pub field_specs: Vec<FieldSpec>,
+    // common blob tables
+    pub signatures: SignaturePool,
+    pub identifiers: IdentifierPool,
+    // constants
+    pub byte_array_pool: U8VectorPool,
+    pub address_pool: AddressPool,
+}
+
+// In `CompiledScript` we replace the FunctionDefinition for main(..) with broken up data
+// to set up the call frame and define/verify main signature and arguments
+pub struct CompiledScript {
+    pub header: CompiledHeader,
+
+    // Below should be enough information to materialize a
+    // main<Cup<T>, W>(arg1, ..., argn) {}
+    // with args of any type (including generic types) that the VM (verification
+    // and runtime) can use.
+    // There is a verification step here that is *purely* runtime.
+    // The client is passing the type arguments and you can imagine restrictions
+    // given our resource semantic.
+
+    // function arguments signature, main is void on return
+    pub arg_types: SignatureIndex,
+    // type parameter for the script
+    pub type_params: Vec<Kind>,
+    // code
+    pub main: CodeUnit,
+}
+
+// A CompiledModule is what it used to be but you will notice the removal of `FieldDefinition`s.
+// Fields are now serialized with the `StructDefinition` which changes verification around fields.
+// (see `StructDefinition` for details)
+pub struct CompiledModule {
+    pub header: CompiledHeader,
+
+    pub struct_defs: Vec<StructDefinition>,
+    pub function_defs: Vec<FunctionDefinition>,
+}
+
+//
+// Basics...
+//
+
+pub enum Kind {
+    All,
+    Resource,
+    Unrestricted,
+}
+
+// Any expressible type in Move
+pub enum Type {
+    Bool,
+    U8,
+    U64,
+    U128,
+    Address,
+    Vector(Box<Type>),
+    Struct(StructHandleIndex),
+    // a complete or partial instantiation of a generic type
+    StructSpec(StructSpecIndex),
+    Reference(Box<Type>),
+    MutableReference(Box<Type>),
+    // a type argument provided by the definition (via instantiation)
+    TypeParameter(TypeParameterIndex),
+}
+
+//
+// Handles
+//
+
+
+pub struct ModuleHandle {
+    pub address: AddressPoolIndex,
+    pub name: IdentifierIndex,
+}
+
+
+pub struct StructHandle {
+    pub module: ModuleHandleIndex,
+    pub name: IdentifierIndex,
+    pub is_nominal_resource: bool,
+    pub type_params: Vec<Kind>,
+}
+
+// FunctionHandles inlines the signature as indexes into the sig table
+pub struct FunctionHandle {
+    pub module: ModuleHandleIndex,
+    pub name: IdentifierIndex,
+    pub arg_types: SignatureIndex,
+    pub return_types: SignatureIndex,
+    pub type_params: Vec<Kind>,
+}
+
+// A field access info (owner type and offset)
+pub struct FieldHandle {
+    owner: StructDefinitionIndex,
+    field: FieldOffset,
+}
+
+//
+// Specifications
```

Comment:
Alternative name ideas: 
- `Binding` (completely or partially "binding" a handle to its instantiation)
- `Instantiation` (since you use that term elsewhere and it seems descriptive. we could then call the current `instantiation` fields `signature`, `types`, or `type_actuals`)

My vote would be for `Specification` -> `Instantiation` + `instantiation` -> `types`, but I don't have a strong opinion. As mentioned before, I do think it's important to use a name other than `Spec` because the prover has a concept of both struct specs and function specs that would be hard to rename to something else.

Reply:
let me try

Reply:
.. and take a look...

Reply:
I am liking how `Instantiation` looks